src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs

```diff
@@ -881,4 +883,152 @@ protected static async Task<ImmutableArray<string>> GetAllMatchingGlobalAliasNam
 
         return result.ToImmutableAndClear();
     }
+
+    /// <summary>
+    /// Discover implied symbols, which are defined based on the requested symbol.
+    /// Implied symbols include anonymous type properties and tuple fields.
+    /// </summary>
+    protected static async ValueTask DiscoverImpliedSymbolsAsync(
+        TSymbol symbol, Solution solution, ArrayBuilder<ISymbol> symbolBuilder, CancellationToken cancellationToken)
+    {
+        using var _ = PooledHashSet<ISymbol>.GetInstance(out var symbolSet);
+
+        await ProducerConsumer<Document>.RunParallelAsync(
+            GetCandidateDocumentsAsync(),
+            produceItems: static async (document, callback, args, cancellationToken) =>
+            {
+                var (symbol, symbolSet) = args;
+                var name = symbol.Name;
+
+                var shouldProcess = await ShouldProcessDocumentAsync(document, name, cancellationToken).ConfigureAwait(false);
+                if (shouldProcess)
+                {
+                    callback(document);
+                }
+            },
+            consumeItems: static async (documents, args, cancellationToken) =>
```

Comment:
I thought consumeItems could be called concurrently. So this isn't safe with your hash set